eclipse-theia/theia
# Use file service in preference provider initialization

#### What it does


Fixes #8993 (partially) by ensuring that file-reading preference providers read the contents of their assigned files before declaring themselves ready. Ready state propagates through the scope providers to the `PreferenceService`, so that when the `PreferenceService.ready` promise resolves, all of its delegated providers will have read the contents of the files assigned to them. This still means that a `.get` call in a `@postConstruct` method that does *not* wait for `.ready` promise can get bad data, but the `.ready` is now reliable. 

This PR doesn't go as far as the functionality suggested in #7426 because it still leaves the monaco text model creation logic in the `@postConstruct` method, but it moves closer.

#### How to test


1. I've written a test that shows in that tick after `ready` resolves, the expected preference values are available.
2. Add a preference to your user-scoped `package.json` that differs from the default. E.g. `"editor.fontSize" : 24`, and make sure there is no conflicting specification in workspace or folder scopes of the workspace you intend to open.
3. In a contribution with a synchronous `@postConstruct` method that injects the `PreferenceService` (e.g. `packages/editor/src/browser/editor-navigation-contribution.ts`), add the following lines to the `.init()` method:
```
console.log('SENTINEL FOR THE VALUE IMMEDIATELY', this.preferenceService.get('editor.fontSize'));
        this.preferenceService.ready.then(() => console.log('SENTINEL FOR THE VALUE WHEN READY RESOLVES', this.preferenceService.get('editor.fontSize')));
        setTimeout(() => console.log('SENTINEL FOR THE VALUE AFTER FIVE SECONDS', this.preferenceService.get('editor.fontSize')), 5000);
```
4. Build and start the application.
5. The `VALUE IMMEDIATELY` log will probably have the default (`14`, in this case); the other two should both show the configured value (`20`, e.g.)

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Colin Grant <colin.grant@ericsson.com>
